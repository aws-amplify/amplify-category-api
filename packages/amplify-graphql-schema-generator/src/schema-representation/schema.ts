import { Model } from './types';
import { Engine } from './engine';

export class Schema {
  private models: Model[] = [];
  constructor(private engine: Engine) {}

  public getModels(): Model[] {
    return this.models;
  }

  public getEngine(): Engine {
    return this.engine;
  }

  public addModel(model: Model): void {
    const modelName = model.getName();
    if (this.hasModel(modelName)) {
      throw new Error(`Model "${modelName}" already exists`);
    }
    this.models.push(model);
  }

  public hasModel(name: string): boolean {
    return this.models.some((model) => model.getName() === name);
  }
}
