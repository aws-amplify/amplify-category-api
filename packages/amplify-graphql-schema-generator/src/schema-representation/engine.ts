export type DBEngineType = 'MySQL' | 'Postgres' | 'DynamoDB';

export class Engine {
  constructor(public type: DBEngineType) {}
}
