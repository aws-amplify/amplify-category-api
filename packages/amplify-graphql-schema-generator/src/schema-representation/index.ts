export * from './engine';
export * from './schema';
export * from './types';
