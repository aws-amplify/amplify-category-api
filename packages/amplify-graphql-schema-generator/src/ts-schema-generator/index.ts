export * from './generate-ts-schema';
export * from './ts-schema-generator';
export { DataSourceGenerateConfig } from './helpers';
