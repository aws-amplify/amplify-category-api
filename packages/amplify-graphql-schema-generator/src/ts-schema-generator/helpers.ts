import ts from 'typescript';
import { TYPESCRIPT_DATA_SCHEMA_CONSTANTS } from 'graphql-transformer-common';
import { VpcConfig } from '@aws-amplify/graphql-transformer-interfaces';
import { DBEngineType, Field, FieldType, Model, Schema } from '../schema-representation';

const GQL_TYPESCRIPT_DATA_SCHEMA_TYPE_MAP = {
  string: 'string',
  int: 'integer',
  float: 'float',
  boolean: 'boolean',
  id: 'id',
  awsdate: 'date',
  awstime: 'time',
  awsdatetime: 'datetime',
  awstimestamp: 'timestamp',
  awsjson: 'json',
  awsemail: 'email',
  awsphone: 'phone',
  awsurl: 'url',
  awsipaddress: 'ipAddress',
};

/**
 * Creates a typescript data schema property from internal SQL schema representation
 * Example typescript data schema property output: `id: a.string().required()`
 * @param field SQL IR field
 * @returns Typescript data schema property in TS Node format
 */
const createProperty = (field: Field): ts.Node => {
  const typeExpression = createDataType(field.type);
  return ts.factory.createPropertyAssignment(ts.factory.createIdentifier(field.name), typeExpression as ts.Expression);
};

/**
 * Creates a typescript data schema type from internal SQL schema representation
 * Example typescript data schema type output: `a.string().required()`
 * @param type SQL IR field type
 * @returns Typescript data schema type in TS Node format
 */
const createDataType = (type: FieldType): ts.Node => {
  if (type.kind === 'Scalar') {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(`${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A}.${getTypescriptDataSchemaType(type.name)}`),
      undefined,
      undefined,
    );
  }

  if (type.kind === 'Enum') {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(`${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A}.${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.ENUM_METHOD}`),
      undefined,
      [
        ts.factory.createArrayLiteralExpression(
          type.values.map((value) => ts.factory.createStringLiteral(value)),
          true,
        ),
      ],
    );
  }

  // We do not import any Database type as 'Custom' type.
  // In case if there is a custom type in the IR schema, we will import it as string.
  if (type.kind === 'Custom') {
    return ts.factory.createCallExpression(
      ts.factory.createIdentifier(`${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A}.${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.STRING_METHOD}`),
      undefined,
      undefined,
    );
  }

  // List or NonNull
  const modifier = type.kind === 'List' ? TYPESCRIPT_DATA_SCHEMA_CONSTANTS.ARRAY_METHOD : TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REQUIRED_METHOD;
  return ts.factory.createCallExpression(
    ts.factory.createPropertyAccessExpression(createDataType(type.type) as ts.Expression, ts.factory.createIdentifier(modifier)),
    undefined,
    undefined,
  );
};

const getTypescriptDataSchemaType = (type: string): string => {
  const DEFAULT_DATATYPE = TYPESCRIPT_DATA_SCHEMA_CONSTANTS.STRING_METHOD;
  const tsDataSchemaType = GQL_TYPESCRIPT_DATA_SCHEMA_TYPE_MAP[type.toLowerCase()];
  return tsDataSchemaType ?? DEFAULT_DATATYPE;
};

const createModelDefinition = (model: Model): ts.Node => {
  const properties = model.getFields().map((field) => {
    return createProperty(field);
  });
  return ts.factory.createObjectLiteralExpression(properties as ts.ObjectLiteralElementLike[], true);
};

const createModel = (model: Model): ts.Node => {
  const modelExpr = ts.factory.createCallExpression(
    ts.factory.createIdentifier(`${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A}.${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.MODEL_METHOD}`),
    undefined,
    [createModelDefinition(model) as ts.Expression],
  );

  // Add primary key to the model
  const modelExprWithPK = ts.factory.createCallExpression(
    ts.factory.createPropertyAccessExpression(modelExpr, ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.IDENTIFIER_METHOD)),
    undefined,
    [
      ts.factory.createArrayLiteralExpression(
        model
          .getPrimaryKey()
          .getFields()
          .map((field) => ts.factory.createStringLiteral(field)),
        true,
      ),
    ],
  );

  return ts.factory.createPropertyAssignment(ts.factory.createStringLiteral(model.getName()), modelExprWithPK as ts.Expression);
};

const exportModifier = ts.factory.createModifier(ts.SyntaxKind.ExportKeyword);

export type DataSourceGenerateConfig = {
  secretNames: {
    connectionUri: string;
    sslCertificate?: string;
  };
  identifier: string;
  vpcConfig?: VpcConfig;
};

/**
 * Creates a typescript data schema from internal SQL schema representation
 * Example typescript data schema:
 * ```
 * export const schema = a.schema({              <--- createSchema()
 *   User: a.model({                             <--- createModel()
 *     id: a.string().required(),                <--- createProperty()
 *     name: a.string(),
 *     status: a.enum(['ACTIVE', 'INACTIVE']),
 *   }),
 * });
 * @param schema
 * @returns Typescript data schema in TS Node format
 */
export const createSchema = (schema: Schema, config?: DataSourceGenerateConfig): ts.Node => {
  const modelsWithPrimaryKeyCount = schema.getModels().filter((model) => model.getPrimaryKey()).length;
  if (modelsWithPrimaryKeyCount === 0) {
    throw new Error('No valid tables found. Make sure at least one table has a primary key.');
  }

  const models = schema
    .getModels()
    .filter((model) => model.getPrimaryKey())
    .map((model) => {
      return createModel(model);
    });
  const tsSchema = ts.factory.createCallExpression(
    ts.factory.createPropertyAccessExpression(
      createConfigureExpression(schema, config),
      ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.SCHEMA_METHOD),
    ),
    undefined,
    [ts.factory.createObjectLiteralExpression(models as ts.ObjectLiteralElementLike[], true)],
  );
  return ts.factory.createVariableStatement(
    [exportModifier],
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.EXPORT_VARIABLE_NAME),
          undefined,
          undefined,
          tsSchema,
        ),
      ],
      ts.NodeFlags.Const,
    ),
  );
};

/**
 * Creates an import expression for typescript data schema
 * @returns Import statement in TS Node format
 */
export const createImportExpression = (containsSecret: boolean): ts.NodeArray<ts.ImportDeclaration> => {
  const importStatement = ts.factory.createImportDeclaration(
    undefined,
    ts.factory.createImportClause(
      false,
      undefined,
      ts.factory.createNamedImports([
        ts.factory.createImportSpecifier(false, undefined, ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A)),
      ]),
    ),
    ts.factory.createStringLiteral(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.SCHEMA_PACKAGE),
  );
  const internalsConfigureImportStatement = ts.factory.createImportDeclaration(
    undefined,
    ts.factory.createImportClause(
      false,
      undefined,
      ts.factory.createNamedImports([
        ts.factory.createImportSpecifier(
          false,
          undefined,
          ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.INTERNALS_CONFIGURE_METHOD),
        ),
      ]),
    ),
    ts.factory.createStringLiteral(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.SCHEMA_PACKAGE_INTERNALS),
  );
  const secretImportStatement = ts.factory.createImportDeclaration(
    undefined,
    ts.factory.createImportClause(
      false,
      undefined,
      ts.factory.createNamedImports([
        ts.factory.createImportSpecifier(
          false,
          undefined,
          ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.BACKEND_SECRET_METHOD),
        ),
      ]),
    ),
    ts.factory.createStringLiteral(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.BACKEND_PACKAGE),
  );

  const importStatementWithEslintDisabled = ts.addSyntheticLeadingComment(
    importStatement,
    ts.SyntaxKind.MultiLineCommentTrivia,
    ' eslint-disable ',
    true,
  );

  const importStatementWithAutoGeneratedComment = ts.addSyntheticLeadingComment(
    importStatementWithEslintDisabled,
    ts.SyntaxKind.MultiLineCommentTrivia,
    ' THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. ',
    true,
  );

  const importExpressions = [importStatementWithAutoGeneratedComment];
  if (containsSecret) {
    importExpressions.push(internalsConfigureImportStatement, secretImportStatement);
  }
  return ts.factory.createNodeArray(importExpressions);
};

export const createConfigureExpression = (schema: Schema, config: DataSourceGenerateConfig): ts.Expression => {
  if (!config) {
    return ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REFERENCE_A);
  }
  const databaseConfig = [
    ts.factory.createPropertyAssignment(
      ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_IDENTIFIER),
      ts.factory.createStringLiteral(config.identifier),
    ),
    ts.factory.createPropertyAssignment(
      ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_ENGINE),
      ts.factory.createStringLiteral(convertDBEngineToDBProtocol(schema.getEngine().type)),
    ),
  ];

  if (config?.secretNames?.connectionUri) {
    databaseConfig.push(
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_CONNECTION_URI),
        ts.factory.createCallExpression(ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.BACKEND_SECRET_METHOD), undefined, [
          ts.factory.createStringLiteral(config.secretNames.connectionUri),
        ]),
      ),
    );
  }

  if (config?.secretNames?.sslCertificate) {
    databaseConfig.push(
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_SSL_CERTIFICATE),
        ts.factory.createCallExpression(ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.BACKEND_SECRET_METHOD), undefined, [
          ts.factory.createStringLiteral(config.secretNames.sslCertificate),
        ]),
      ),
    );
  }

  if (config.vpcConfig) {
    databaseConfig.push(
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_VPC),
        createVpcConfigExpression(config.vpcConfig),
      ),
    );
  }

  const dbConfig = ts.factory.createObjectLiteralExpression(
    [
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_DATABASE),
        ts.factory.createObjectLiteralExpression(databaseConfig, true),
      ),
    ],
    true,
  );
  const configureExpr = ts.factory.createCallExpression(
    ts.factory.createIdentifier(`${TYPESCRIPT_DATA_SCHEMA_CONSTANTS.INTERNALS_CONFIGURE_METHOD}`),
    undefined,
    [dbConfig],
  );
  return configureExpr;
};

const convertDBEngineToDBProtocol = (engine: DBEngineType): 'mysql' | 'postgresql' | 'dynamodb' => {
  switch (engine) {
    case 'MySQL':
      return 'mysql';
    case 'Postgres':
      return 'postgresql';
    default:
      return 'dynamodb';
  }
};

const createVpcConfigExpression = (vpc: VpcConfig): ts.Expression => {
  const vpcIdExpression = ts.factory.createPropertyAssignment(
    ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_VPC_ID),
    ts.factory.createStringLiteral(vpc.vpcId),
  );
  const securityGroupIdsExpression = ts.factory.createPropertyAssignment(
    ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_SECURITY_GROUP_IDS),
    ts.factory.createArrayLiteralExpression(
      vpc.securityGroupIds.map((sg) => ts.factory.createStringLiteral(sg)),
      true,
    ),
  );
  const subnetAvailabilityZoneExpression = ts.factory.createPropertyAssignment(
    ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_AZ_CONFIG),
    ts.factory.createArrayLiteralExpression(
      vpc.subnetAvailabilityZoneConfig.map((az) => {
        return ts.factory.createObjectLiteralExpression(
          [
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_SUBNET_ID),
              ts.factory.createStringLiteral(az.subnetId),
            ),
            ts.factory.createPropertyAssignment(
              ts.factory.createIdentifier(TYPESCRIPT_DATA_SCHEMA_CONSTANTS.PROPERTY_AZ),
              ts.factory.createStringLiteral(az.availabilityZone),
            ),
          ],
          true,
        );
      }),
      true,
    ),
  );

  const vpcConfig = ts.factory.createObjectLiteralExpression(
    [vpcIdExpression, securityGroupIdsExpression, subnetAvailabilityZoneExpression],
    true,
  );

  return vpcConfig;
};
