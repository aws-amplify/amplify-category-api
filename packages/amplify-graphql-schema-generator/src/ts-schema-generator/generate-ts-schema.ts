import * as ts from 'typescript';
import { Schema } from '../schema-representation';
import { createImportExpression, createSchema, DataSourceConfig } from './helpers';

const file = ts.createSourceFile('schema.ts', '', ts.ScriptTarget.ESNext, false, ts.ScriptKind.TS);
const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

export const generateTypescriptDataSchema = (schema: Schema, config?: DataSourceConfig): string => {
  const containsSecretName = !!config?.secretName;
  const result = printer.printList(
    ts.ListFormat.MultiLine,
    ts.factory.createNodeArray([
      ...createImportExpression(containsSecretName),
      ts.factory.createIdentifier('\n'),
      createSchema(schema, config),
    ]),
    file,
  );
  return result;
};
