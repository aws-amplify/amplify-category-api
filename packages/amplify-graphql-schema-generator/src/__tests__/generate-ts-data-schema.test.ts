import { Engine, Field, Model, Schema } from '../schema-representation';
import { generateTypescriptDataSchema } from '../ts-schema-generator/generate-ts-schema';
import { DataSourceConfig } from '../ts-schema-generator/helpers';
import { TypescriptDataSchemaGenerator } from '../ts-schema-generator/ts-schema-generator';

jest.mock('../utils', () => ({
  getHostVpc: jest.fn(() => {
    return {
      vpcId: 'abc',
      securityGroupIds: ['sg0', 'sg1', 'sg2'],
      subnetAvailabilityZoneConfig: [
        {
          subnetId: 'sb0',
          availabilityZone: 'az0',
        },
        {
          subnetId: 'sb1',
          availabilityZone: 'az1',
        },
        {
          subnetId: 'sb2',
          availabilityZone: 'az2',
        },
      ],
    };
  }),
}));

describe('Type name conversions', () => {
  it('ts schema generator should invoke generate schema', async () => {
    const dbschema = new Schema(new Engine('MySQL'));
    const model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const buildSchemaMockMethod = jest.spyOn(TypescriptDataSchemaGenerator as any, 'buildSchema');
    buildSchemaMockMethod.mockImplementation(() => {
      return dbschema;
    });

    const schema = await TypescriptDataSchemaGenerator.generate({
      engine: 'mysql',
      host: 'host',
      port: 3306,
      database: 'database',
      username: 'username',
      password: 'password',
      connectionUriSecretName: 'secret',
    });
    expect(schema).toMatchSnapshot();
  });

  it('basic models should generate correct typescript data schema', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateTypescriptDataSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('check all valid datatypes', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    const model = new Model('Table');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('field1', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('field2', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('field3', { kind: 'Scalar', name: 'Int' }));
    model.addField(new Field('field4', { kind: 'Scalar', name: 'Float' }));
    model.addField(new Field('field5', { kind: 'Scalar', name: 'Boolean' }));
    model.addField(new Field('field6', { kind: 'Scalar', name: 'ID' }));
    model.addField(new Field('field7', { kind: 'Scalar', name: 'AWSDate' }));
    model.addField(new Field('field8', { kind: 'Scalar', name: 'AWSTime' }));
    model.addField(new Field('field9', { kind: 'Scalar', name: 'AWSDateTime' }));
    model.addField(new Field('field10', { kind: 'Scalar', name: 'AWSTimestamp' }));
    model.addField(new Field('field11', { kind: 'Scalar', name: 'AWSJSON' }));
    model.addField(new Field('field12', { kind: 'Scalar', name: 'AWSEmail' }));
    model.addField(new Field('field13', { kind: 'Scalar', name: 'AWSPhone' }));
    model.addField(new Field('field14', { kind: 'Scalar', name: 'AWSURL' }));
    model.addField(new Field('field15', { kind: 'Scalar', name: 'AWSIPAddress' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateTypescriptDataSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generates enum imports correctly', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE', 'INACTIVE'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateTypescriptDataSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('schema with database config secret and vpc should generate typescript data schema with configure', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const config: DataSourceConfig = {
      identifier: 'ID1234567890',
      secretName: 'CONN_STR',
      vpcConfig: {
        vpcId: '123',
        securityGroupIds: ['sb1', 'sb2', 'sb3'],
        subnetAvailabilityZoneConfig: [
          {
            subnetId: 'sb1',
            availabilityZone: 'az1',
          },
          {
            subnetId: 'sb2',
            availabilityZone: 'az2',
          },
          {
            subnetId: 'sb3',
            availabilityZone: 'az3',
          },
        ],
      },
    };

    const graphqlSchema = generateTypescriptDataSchema(dbschema, config);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('postgres schema with database config secret and vpc should generate typescript data schema with configure', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const config: DataSourceConfig = {
      identifier: 'ID1234567890',
      secretName: 'CONN_STR',
      vpcConfig: {
        vpcId: '123',
        securityGroupIds: ['sb1'],
        subnetAvailabilityZoneConfig: [
          {
            subnetId: 'sb1',
            availabilityZone: 'az1',
          },
        ],
      },
    };

    const graphqlSchema = generateTypescriptDataSchema(dbschema, config);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('schema with database config without vpc should generate typescript data schema with configure', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const config: DataSourceConfig = {
      identifier: 'ID1234567890',
      secretName: 'CONN_STR',
    };

    const graphqlSchema = generateTypescriptDataSchema(dbschema, config);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generate typescript data schema should skip processing tables without primary key', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    dbschema.addModel(model);

    const config: DataSourceConfig = {
      identifier: 'ID1234567890',
      secretName: 'CONN_STR',
    };

    const graphqlSchema = generateTypescriptDataSchema(dbschema, config);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generate typescript data schema throw error if none of the tables have primary key', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    dbschema.addModel(model);

    model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    dbschema.addModel(model);

    const config: DataSourceConfig = {
      identifier: 'ID1234567890',
      secretName: 'CONN_STR',
    };

    expect(() => generateTypescriptDataSchema(dbschema, config)).toThrowError(
      'No valid tables found. Make sure at least one table has a primary key.',
    );
  });
});
