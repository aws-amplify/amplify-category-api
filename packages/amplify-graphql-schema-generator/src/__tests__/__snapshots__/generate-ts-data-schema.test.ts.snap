// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type name conversions basic models should generate correct typescript data schema 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";

export const schema = a.schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ]),
    \\"Profile\\": a.model({
        id: a.string().required(),
        details: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions check all valid datatypes 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";

export const schema = a.schema({
    \\"Table\\": a.model({
        id: a.string().required(),
        field1: a.string(),
        field2: a.string(),
        field3: a.integer(),
        field4: a.float(),
        field5: a.boolean(),
        field6: a.id(),
        field7: a.date(),
        field8: a.time(),
        field9: a.datetime(),
        field10: a.timestamp(),
        field11: a.json(),
        field12: a.email(),
        field13: a.phone(),
        field14: a.url(),
        field15: a.ipAddress()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions generate typescript data schema should skip processing tables without primary key 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";
import { configure } from \\"@aws-amplify/data-schema/internals\\";
import { secret } from \\"@aws-amplify/backend\\";

export const schema = configure({
    database: {
        identifier: \\"ID1234567890\\",
        engine: \\"mysql\\",
        connectionUri: secret(\\"CONN_STR\\")
    }
}).schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions generates enum imports correctly 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";

export const schema = a.schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string(),
        status: a.enum([
            \\"ACTIVE\\",
            \\"INACTIVE\\"
        ])
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions postgres schema with database config secret and vpc should generate typescript data schema with configure 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";
import { configure } from \\"@aws-amplify/data-schema/internals\\";
import { secret } from \\"@aws-amplify/backend\\";

export const schema = configure({
    database: {
        identifier: \\"ID1234567890\\",
        engine: \\"postgresql\\",
        connectionUri: secret(\\"CONN_STR\\"),
        vpcConfig: {
            vpcId: \\"123\\",
            securityGroupIds: [
                \\"sb1\\"
            ],
            subnetAvailabilityZones: [
                {
                    subnetId: \\"sb1\\",
                    availabilityZone: \\"az1\\"
                }
            ]
        }
    }
}).schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions schema with database config secret and vpc should generate typescript data schema with configure 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";
import { configure } from \\"@aws-amplify/data-schema/internals\\";
import { secret } from \\"@aws-amplify/backend\\";

export const schema = configure({
    database: {
        identifier: \\"ID1234567890\\",
        engine: \\"mysql\\",
        connectionUri: secret(\\"CONN_STR\\"),
        vpcConfig: {
            vpcId: \\"123\\",
            securityGroupIds: [
                \\"sb1\\",
                \\"sb2\\",
                \\"sb3\\"
            ],
            subnetAvailabilityZones: [
                {
                    subnetId: \\"sb1\\",
                    availabilityZone: \\"az1\\"
                },
                {
                    subnetId: \\"sb2\\",
                    availabilityZone: \\"az2\\"
                },
                {
                    subnetId: \\"sb3\\",
                    availabilityZone: \\"az3\\"
                }
            ]
        }
    }
}).schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ]),
    \\"Profile\\": a.model({
        id: a.string().required(),
        details: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions schema with database config without vpc should generate typescript data schema with configure 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";
import { configure } from \\"@aws-amplify/data-schema/internals\\";
import { secret } from \\"@aws-amplify/backend\\";

export const schema = configure({
    database: {
        identifier: \\"ID1234567890\\",
        engine: \\"mysql\\",
        connectionUri: secret(\\"CONN_STR\\")
    }
}).schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ]),
    \\"Profile\\": a.model({
        id: a.string().required(),
        details: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;

exports[`Type name conversions ts schema generator should invoke generate schema 1`] = `
"/* eslint-disable */
/* THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. */
import { a } from \\"@aws-amplify/data-schema\\";
import { configure } from \\"@aws-amplify/data-schema/internals\\";
import { secret } from \\"@aws-amplify/backend\\";

export const schema = configure({
    database: {
        identifier: \\"IDNb9LVUQdGlNwNzYAgL6A\\",
        engine: \\"mysql\\",
        connectionUri: secret(\\"secret\\"),
        vpcConfig: {
            vpcId: \\"abc\\",
            securityGroupIds: [
                \\"sg0\\",
                \\"sg1\\",
                \\"sg2\\"
            ],
            subnetAvailabilityZones: [
                {
                    subnetId: \\"sb0\\",
                    availabilityZone: \\"az0\\"
                },
                {
                    subnetId: \\"sb1\\",
                    availabilityZone: \\"az1\\"
                },
                {
                    subnetId: \\"sb2\\",
                    availabilityZone: \\"az2\\"
                }
            ]
        }
    }
}).schema({
    \\"User\\": a.model({
        id: a.string().required(),
        name: a.string()
    }).identifier([
        \\"id\\"
    ])
});
"
`;
